### PR TITLE
 nixos/tailscale-derper: init 

### PR DESCRIPTION
**Pull request by SuperSandro2000**
## Description of changes

inlcudes #306532

when I am done figuring out how to setup derp completely, I will probably do a PR for a module.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by danderson — COMMENTED** [review]
Looks reasonable to me, @ me when you want a review of the nixos module? If you'd like to merge the packaging change and do the module separately, that's also fine by me.

I assume you've found it already, but https://tailscale.com/kb/1118/custom-derp-servers has some orientation on how to run derper. It may also be worth linking to from the module docs?

Also worth noting that tailscale doesn't provide any support for running custom derp, unless you're paying enterprise kind of money. I'm not sure if that's worth mentioning in the nixos module or not, I can't tell how many people would use this and how many would try to file support tickets. I'm ok with it either way.

**Comment by SuperSandro2000** [comment]
> Also worth noting that tailscale doesn't provide any support for running custom derp, unless you're paying enterprise kind of money. I'm not sure if that's worth mentioning in the nixos module or not, I can't tell how many people would use this and how many would try to file support tickets. I'm ok with it either way.

It works with headscale :)

**Comment by SuperSandro2000** [comment]
@danderson I've added a module. I've extracted it from my config and I think I didn't miss anything.

**Comment by SuperSandro2000** [comment]
> I assume you've found it already, but [tailscale.com/kb/1118/custom-derp-servers](https://tailscale.com/kb/1118/custom-derp-servers) has some orientation on how to run derper. It may also be worth linking to from the module docs?

Is that page hidden from search engines?

---

I've been running the setup for a bit now and I think it is ready and stable and works good.

File: nixos/modules/services/networking/tailscale-derper.nix [hunk 1]
```diff
@@ -0,0 +1,114 @@
+{ config, lib, pkgs, ... }:
```
Comment (fpletz) [inline nixos/modules/services/networking/tailscale-derper.nix:1]: Needs some nixfmt-rfc-style. :smile_cat:

File: nixos/modules/services/networking/tailscale-derper.nix [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+          STUN port to listen on.
+          See online docs <https://tailscale.com/kb/1118/custom-derp-servers#prerequisites> how to configure a different port.
+        '';
+      };
+
+      verifyClients = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = ''
+          Wether to verify clients against a locally running tailscale daemon if they are allowed to connect to this node or not.
```
Comment (h7x4) [inline nixos/modules/services/networking/tailscale-derper.nix:50]:
```suggestion
          Whether to verify clients against a locally running tailscale daemon if they are allowed to connect to this node or not.
```

File: nixos/modules/services/networking/tailscale-derper.nix [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    networking.firewall = lib.mkIf cfg.openFirewall {
+      allowedTCPPorts = [ 80 ];
+      allowedUDPPorts = [
+        443
+        cfg.port
```
Comment (h7x4) [inline nixos/modules/services/networking/tailscale-derper.nix:61]: I don't know this service too well, just asking to verify that it is necessary to expose both `443` and `cfg.port` when `443` seems to reverse proxy to `cfg.port`. Is it a case of using http as an alternative entrypoint when `cfg.port` doesn't work?

File: nixos/modules/services/networking/tailscale-derper.nix [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        enable = true;
+        upstreams.tailscale-derper = {
+          servers."127.0.0.1:${toString cfg.port}" = { };
+          extraConfig = ''
+            keepalive 64;
+          '';
+        };
+        virtualHosts."${cfg.domain}" = {
+          addSSL = true;
+          enableACME = true;
```
Comment (h7x4) [inline nixos/modules/services/networking/tailscale-derper.nix:76]: I suppose there might be cases where people want to manage certificates on their own.

```suggestion
          addSSL = lib.mkDefault true;
          enableACME = lib.mkDefault true;
```

**Review by h7x4 — APPROVED** [review]

File: nixos/modules/services/networking/tailscale-derper.nix (hunk 4 above)
Comment (SuperSandro2000) [reply nixos/modules/services/networking/tailscale-derper.nix:76]: addSSL is necessary, ACME can indeed be done via acmehost.

File: nixos/modules/services/networking/tailscale-derper.nix (hunk 3 above)
Comment (SuperSandro2000) [reply nixos/modules/services/networking/tailscale-derper.nix:61]: actually port 443/udp is not required.

File: nixos/modules/services/networking/tailscale-derper.nix [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        description = ''
+          STUN port to listen on.
+          See online docs <https://tailscale.com/kb/1118/custom-derp-servers#prerequisites> on how to configure a different external port.
+        '';
+      };
+
+      port = lib.mkOption {
+        type = lib.types.port;
+        default = 8010;
+        description = "The port the derper process will listen on. This is not the port tailscale will connect to.";
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/tailscale-derper.nix:49]: To keep things simple, I decided to keep the default port 80 and 443 tcp for now.

**Comment by ocfox** [comment]
Used weeks, work fine!

**Review by fpletz — APPROVED** [review]

**Comment by adamcstephens** [comment]
I appreciate things working out of the box, but it would be preferable to wrap optional conditionals around the default use of a proxy. Not everyone uses nginx, and a warning about a missing cert for nginx is confusing if you're not using it.

**Pull request merged**